feat(internal): Pass 8 CSPRNG random integer words

craciunoiuc opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [ ] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

craciunoiuc committed:
feat(internal): Pass 8 CSPRNG random integer words

Depends-on: unikraft/unikraft#1496

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(unikraft): Add ukrandom library parameters

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung commented on unikraft/export/v0/ukrandom/params.go:
```diff
@@ -0,0 +1,54 @@
+	"kraftkit.sh/unikraft/export/v0/ukargparse"
+)
+
+var ParamRandomBytes = ukargparse.NewParamStrSlice("random", "seed", nil)
```
Let's call this var, to reflect its name in the cmdline, just `ParamRandomSeed`.

Subsequent changes are then necessary to the constructor.

nderjung reviewed (approved):
All good on my side implementation-wise. Thanks!

Regarding the interaction with Unikraft itself, I'd like to cc @razvand and @michpappas before merging to help verify everything in conjunction with https://github.com/unikraft/unikraft/pull/1496

craciunoiuc commented:
Updated to check for RDRAND/RDSEED

If they are both present then ignore the config option and let Unikraft take its own bytes from the device/generator

michpappas commented on unikraft/export/v0/ukrandom/params.go:
```diff
@@ -0,0 +1,54 @@
+
+	sb.WriteString("[ ")
+	for i := 0; i < 8; i++ {
+		sb.WriteString(fmt.Sprintf("%04x ", rng[i]))
```
You need to prepend each number with `0x` otherwise they won't be interpreted by `uklibparam` neither as valid hex, nor as decimals.

michpappas commented on internal/cli/kraft/run/runner_kraftfile_runtime.go:
```diff
@@ -358,6 +360,16 @@ func (runner *runnerKraftfileRuntime) Prepare(ctx context.Context, opts *RunOpti
+	if !runtime.KConfig().AllNoOrUnset(
+		"CONFIG_LIBUKRANDOM",
+		"CONFIG_LIBUKRANDOM_CMDLINE_INIT",
+	) && !(cpuid.CPU.Rdrand() && cpuid.CPU.Rdseed()) {
```
There is a theoretical case where a CPU has `RDSEED` but not `RDRAND`, but I haven't come across such a thing in practice, so I'm not sure if it's worth adding a check. Saying that ,the opposite is possible: `RDRAND` was introduced before `RDSEED`, so there are CPUs out there that have the former but not the latter. The implementation in Unikraft has the minimum requirement of  `RDRAND` when the driver is enabled (seeding functions fall back to `RDRAND`), so in theory you may skip the check for `RDSEED` altogether.

Another thing related to portability: I had a quick look at https://github.com/klauspost/cpuid and it looks like it lacks detection for `FEAT_RNG` on Arm (along with many others). You may consider opening a PR upstream to add these so that they're there by the time we need them.

craciunoiuc replied:
noted, will do

michpappas reviewed (changes requested):
@craciunoiuc just a few comments. Also if you update this, it would be great if you posted a few of the generated cmdlines.

craciunoiuc commented:
```
qemu-system-x86_64 -append random.seed=[ 0xef59fa23 0x1eba1280 0xb0972ca2 0xc734f02f 0x2079dc04 0x158e0490 0x656d5095 0xa407c115 ] -- /usr/bin/nginx  -cpu qemu64,+pdpe1gb,+rdrand,+rdseed,-vmx,-svm -daemonize -device virtio-net-pci,mac=02:b0:b0:84:3e:01,netdev=hostnet0 -device pvpanic -display none -kernel /home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/unikraft/bin/kernel -machine pc -m size=128M -monitor unix:/home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/qemu_mon.sock,server,nowait -name 99876f61-99a0-4072-a2f8-5b6483635e20 -netdev user,id=hostnet0,hostfwd=tcp::8081-:80 -nographic -no-reboot -S -parallel none -pidfile /home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/machine.pid -qmp unix:/home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/qemu_control.sock,server,nowait -qmp unix:/home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/qemu_events.sock,server,nowait -rtc base=utc -serial file:/home/cez/.local/share/kraftkit/runtime/99876f61-99a0-4072-a2f8-5b6483635e20/machine.log -smp cpus=1,threads=1,sockets=1 -vga none
```

this is how it looks for one run, note that quotes don't show up in the logs

craciunoiuc committed:
gomod(deps): Bump 'klauspost/cpuid' to the latest version

Also update API usage.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

michpappas reviewed (approved):
Looks good, thanks a lot!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

craciunoiuc committed:
fix(machine): Add 'rdrand' feature even if qemu < 8.0.0

This was added in qemu 4.0.0 and we refuse to run
when 4.2.0 or lower is used anyway. Unikraft can now
with rdrand only, so this is not a hard requriment.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung reviewed (approved):
Thanks everyone! Looks good from my side. I'll merge this when https://github.com/unikraft/unikraft/pull/1496 is merged into staging.

nderjung commented:
Linking for completeness: https://github.com/klauspost/cpuid/pull/152

michpappas commented on internal/cli/kraft/run/runner_kraftfile_runtime.go:
```diff
@@ -354,6 +356,17 @@ func (runner *runnerKraftfileRuntime) Prepare(ctx context.Context, opts *RunOpti
 	}
 
+	var kernelArgs []string
+	if (!runtime.KConfig().AllNoOrUnset("CONFIG_LIBUKRANDOM_LCPU") || !cpuid.CPU.Has(cpuid.RDRAND)) &&
```
The 1st condition here looks incorrect. It should check if `CONFIG_LIBUKRANDOM_LCPU` is *not* enabled.

michpappas commented on internal/cli/kraft/run/runner_kraftfile_unikraft.go:
```diff
@@ -147,6 +149,12 @@ func (runner *runnerKraftfileUnikraft) Prepare(ctx context.Context, opts *RunOpt
 		"CONFIG_LIBVFSCORE_AUTOMOUNT_CI_EINITRD",
 	)
 
+	noRandom := t.KConfig().AllNoOrUnset("CONFIG_LIBUKRANDOM_LCPU") && cpuid.CPU.Has(cpuid.RDRAND) ||
```
This one looks like different logic altoghether 🤔

craciunoiuc replied:
did some logic toke reduce the number of negations, will simplfy in another way

michpappas commented on internal/cli/kraft/run/runner_package.go:
```diff
@@ -338,6 +340,17 @@ func (runner *runnerPackage) Prepare(ctx context.Context, opts *RunOptions, mach
 	}
 
+	var kernelArgs []string
+	if (!targ.KConfig().AllNoOrUnset("CONFIG_LIBUKRANDOM_LCPU") || !cpuid.CPU.Has(cpuid.RDRAND)) &&
```
This one too should check if `CONFIG_LIBUKRANDOM_LCPU` is *not* enabled.

michpappas commented on machine/firecracker/v1alpha1.go:
```diff
@@ -87,8 +87,8 @@ func (service *machineV1alpha1Service) Create(ctx context.Context, machine *mach
-	if !cpuid.CPU.Rdrand() || !cpuid.CPU.Rdseed() {
-		log.G(ctx).Warn("RDRAND and RDSEED are not supported by the host CPU to be able to run Unikraft v0.17.0 and greater with hardware randomization")
+	if !cpuid.CPU.Has(cpuid.RDRAND) {
+		log.G(ctx).Warn("RDRAND is not supported by the host CPU to be able to run Unikraft v0.17.0 and greater with hardware randomization")
```
Since you're updating this line, the term normally used is "randomness", not "randomization". Or if you want to be more precise, "CPU-generated randomness".

michpappas commented on machine/qemu/v1alpha1.go:
```diff
@@ -438,8 +437,8 @@ func (service *machineV1alpha1Service) Create(ctx context.Context, machine *mach
-			if !cpuid.CPU.Rdrand() || !cpuid.CPU.Rdseed() {
-				log.G(ctx).Warn("RDRAND and RDSEED are not supported by the host CPU, try rerunning with emulation '-W' to be able to run Unikraft v0.17.0 and greater with hardware randomization")
+			if !cpuid.CPU.Has(cpuid.RDRAND) {
+				log.G(ctx).Warn("RDRAND is not supported by the host CPU, try rerunning with emulation '-W' to be able to run Unikraft v0.17.0 and greater with hardware randomization")
```
Same comment about the message, as above.

michpappas reviewed (changes requested):
@craciunoiuc besides the comments below, I think we should have the conditions on for generating the cmdline rewritten in a more clear / readable way, especially given the security-related nature of that code 🙏🏼 

Also since your changes for `FEAT_RNG` were merged upstream, please consider if you should also update the CPUID check for Arm too.

craciunoiuc commented:
Checks are automatically done when checking for the RDRAND features. Imagine it as a an interface with different implementations depending on the OS running the library. So all is good.

michpappas commented on internal/cli/kraft/run/runner_kraftfile_runtime.go:
```diff
@@ -354,6 +356,18 @@ func (runner *runnerKraftfileRuntime) Prepare(ctx context.Context, opts *RunOpti
 
+	var kernelArgs []string
+	hasCmdlineSupport := !runtime.KConfig().AllNoOrUnset(
+		"CONFIG_LIBUKRANDOM",
```
These are merely "I would had done it differently" / in  `unikraft` we do it differently,  so totally up to you to take into account or ignore. I only care about functional correctness, and the current implementation looks good in that aspect :)

1. I would had written this a bit differently, having a separate variable for the presence of `CONFIG_LIBUKRANDOM`, and then do the check for that at the final condition:
```go
if libUKRandom && hasNoCpuRandomnessSupport && hasCmdlineSupport {
```
 The reason is that there there is no stronger correlation between `CONFIG_LIBUKRANDOM` and `CONFIG_LIBUKRANDOM_CMDLINE_INIT` than `CONFIG_LIBUKRANDOM_CMDLINE_LCPU`.
 
2. I would had also moved this logic into a function to avoid copying (and keeping up to date) into three different places.

3. Looking at the individual commits, it seems that some changes are introduced in one commit just to be updated (read corrected) in another, which people may frown up.

craciunoiuc replied:
1. I can do it, sure
2. If I do it I need to do it in other places too, so I'll skip this one 😆 
3. Yeah, I was a bit split on this also, but I think the bump should be done separately as it has significant changes

michpappas commented on machine/firecracker/v1alpha1.go:
```diff
@@ -87,8 +87,8 @@ func (service *machineV1alpha1Service) Create(ctx context.Context, machine *mach
 
-	if !cpuid.CPU.Rdrand() || !cpuid.CPU.Rdseed() {
-		log.G(ctx).Warn("RDRAND and RDSEED are not supported by the host CPU to be able to run Unikraft v0.17.0 and greater with hardware randomization")
+	if !cpuid.CPU.Has(cpuid.RDRAND) {
```
Should these be additionally conditional to `!hasNoCpuRandomnessSupport`?

craciunoiuc replied:
Makes sense

craciunoiuc replied:
Actually, symbols are not accessible in there, I will move the warning closer to that

michpappas reviewed (approved):
Looks good, thanks!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
